### ISSUE TITLE
msm-fb-refresher makes the phone reboot

### ISSUE THREAD

**Issue by z3ntu** [postmarketOS/pmbootstrap#604]
Last lines I got from the serial console:
```
android_work: android_work: sent uevent USB_STATE=CONFIGURED
BUG: sleeping function called from invalid context at /home/user/build/src/android_kernel_fairphone_msm8974-8c73c363d1cd11624ddda9a1ef08b9dc7a5569e4/kernel/mutex.c:85
Bad IRQ904923528
in_atomic(): 1, irqs_disabled(): 128, pid: 331, name: fbsplash
```

Commenting the line `deviceinfo_msm_refresher="true"` "fixes" the issue.

**Comment by z3ntu** [comment]
Complete output from the serial console (I let it reboot a few times)
[screen-pts-4.g550jk-0.log.txt](https://github.com/postmarketOS/pmbootstrap/files/1319137/screen-pts-4.g550jk-0.log.txt)

**Comment by ollieparanoid** [comment]
Random guess: maybe `msm-fb-refresher` does not get killed correctly at the end of the initramfs, before calling `switch_root`? You could debug this by running `sh` just before `switch_root` and checking via telnet if it is still running, if you want to.

**Comment by drebrez** [comment]
@z3ntu you mentioned in the PR that you needed the `while cat modes > mode`  to make it work, that's something different I think because the `msm-fb-refresher` does the `while echo 0,0 > pan` more like for [this](https://github.com/postmarketOS/pmbootstrap/blob/master/aports/device/device-lg-lenok/fb_fix.sh#L5-L8) device. 
See msm-fb-source code [here](https://github.com/AsteroidOS/msm-fb-refresher/blob/master/refresher.c#L41).
In a branch of the osk-sdl development they tried a solution more similar to the `cat modes` thing, see [here](https://github.com/postmarketOS/osk-sdl/blob/617c1cf407160329535d6dbc6a315fd0dc72de70/main.cpp#L109-L118). 
If we need 2 different solution we can consider having a `msm-fb-refresher` variant? what you think @ollieparanoid ?

**Comment by drebrez** [comment]
from the log there is `fbsplash` right after the reboot, so my guess is that `msm-fb-refresher` unlocked somehow the `fbsplash` functionality and the issue is with that, not with the `msm-fb-refresher`.

@z3ntu can you try these steps:

- comment out [these 2 lines in the init_functions.sh](https://github.com/postmarketOS/pmbootstrap/blob/master/aports/main/postmarketos-mkinitfs/init_functions.sh#L292-L293)
- add a call to `loop_forever` right before switching chroot [here](https://github.com/postmarketOS/pmbootstrap/blob/master/aports/main/postmarketos-mkinitfs/init.sh.in#L43)
- add the usb-shell (`pmbootstrap initfs hook_add usb-shell`)
- flash/boot the new kernel
- start the device
- connect to it with `telnet 172.16.42.1 24`
- try with a simple `cat /dev/urandom > /dev/fb0`

Does it shows something on the screen or it reboots?

**Comment by z3ntu** [comment]
Starting `msm-fb-refresher --loop` while in weston works and refreshes the screen (as expected).

@drebrez I will try that now

**Comment by z3ntu** [comment]
```
/ # cat /dev/urandom > /dev/fb0
cat: write error: No such device
/ # ls -al /dev/urandom /dev/fb0
crw-rw----    1 0        0          29,   0 Jan  1 09:51 /dev/fb0
crw-rw----    1 0        0           1,   9 Sep 21  2017 /dev/urandom
```

**Comment by z3ntu** [comment]
git diff:

```
diff --git a/aports/device/device-fairphone-fp2/APKBUILD b/aports/device/device-fairphone-fp2/APKBUILD
index c743b9d..340ff2c 100644
--- a/aports/device/device-fairphone-fp2/APKBUILD
+++ b/aports/device/device-fairphone-fp2/APKBUILD
@@ -5,7 +5,7 @@ pkgdesc="Fairphone 2"
 url="https://github.com/postmarketOS"
 arch="noarch"
 license="MIT"
-depends="linux-fairphone-fp2 linux-firmware mkbootimg"
+depends="linux-fairphone-fp2 linux-firmware mkbootimg msm-fb-refresher"
 makedepends=""
 install=""
 subpackages=""
diff --git a/aports/main/postmarketos-mkinitfs/APKBUILD b/aports/main/postmarketos-mkinitfs/APKBUILD
index bc14b89..8e40f94 100644
--- a/aports/main/postmarketos-mkinitfs/APKBUILD
+++ b/aports/main/postmarketos-mkinitfs/APKBUILD
@@ -25,6 +25,6 @@ package() {
                "$pkgdir/sbin/mkinitfs"
        mkdir -p "$pkgdir/etc/postmarketos-mkinitfs/hooks/"
 }
-sha512sums="91860e61cc42b77c02522ab273cf4793b78ee595a75c026c057522a5e6d1fdcc7ec868cbdaea5912c90dea61483468dc609dc69144a6d8e68ef47c9d1459fdb6  init.sh.in
-8823d0685c40f4a21936a3a371914a9af39f12ad48371ccdaea88379e2109bdd13728252ce774cc53f6053cecd51334ee3b7a3d559ba4b77e45d0e233714d930  init_functions.sh
+sha512sums="3e7391e46c727ecfcbc1a56edda7ec296e5e498f987544394543196800960ccf1963ceec4e3956a70b09f49978a520ac6c96dbcab12bd3ef30de2f52052721a9  init.sh.in
+609e0f2735d460043b6f621e3a40a68e45fa4a5252a0b324ee61a2e3b2b0ba9198b18ddb4e29e1586f76f25a2f4c5d446a8cf348d8fc80f75dbdda8f7733eb02  init_functions.sh
 8aae74c95df0f0c1cff317d0038c897d7406f29622d9657115b6cab02c6e79ec00b66df8270a1dc62f50ff65483cca96e30810915b24a70fb643ac5a101d65d2  mkinitfs.sh"
diff --git a/aports/main/postmarketos-mkinitfs/init.sh.in b/aports/main/postmarketos-mkinitfs/init.sh.in
index 44eab36..1e298d9 100644
--- a/aports/main/postmarketos-mkinitfs/init.sh.in
+++ b/aports/main/postmarketos-mkinitfs/init.sh.in
@@ -40,6 +40,7 @@ unlock_root_partition
 resize_root_filesystem
 mount_root_partition
 
+loop_forever
 # Switch root
 killall telnetd mdev msm-fb-refresher 2>/dev/null
 umount /boot
diff --git a/aports/main/postmarketos-mkinitfs/init_functions.sh b/aports/main/postmarketos-mkinitfs/init_functions.sh
index 21fa3cf..bdbc19e 100644
--- a/aports/main/postmarketos-mkinitfs/init_functions.sh
+++ b/aports/main/postmarketos-mkinitfs/init_functions.sh
@@ -289,8 +289,9 @@ start_usb_unlock() {
 
 # $1: path to ppm.gz file
 show_splash() {
-       gzip -c -d "$1" >/tmp/splash.ppm
-       fbsplash -s /tmp/splash.ppm
+    echo ""
+#      gzip -c -d "$1" >/tmp/splash.ppm
+#      fbsplash -s /tmp/splash.ppm
 }
 
 echo_connect_ssh_message() {
```

**Comment by z3ntu** [comment]
Okay shell redirection to /dev/fb0 didn't work but I used fbsplash for showing. 
```
/ # ps aux | grep msm-fb-refresher
  255 0          0:00 /usr/sbin/msm-fb-refresher --loop
  420 0          0:00 grep msm-fb-refresher
/ # gzip -c -d /splash-loading.ppm.gz > /tmp/splash.ppm
/ # fbsplash -s /tmp/splash.ppm
/ # # waited for long enough, worked fine
/ # gzip -c -d /splash-telnet.ppm.gz > /tmp/splash.ppm 
/ # fbsplash -s /tmp/splash.ppm
/ # # telnet hanged right there and about 3-5 seconds later the phone rebooted
```

So apparently the kernel doesn't like it when while msm-fb-refresher is running, fbsplash displays a new image

**Comment by drebrez** [comment]
@z3ntu seems like the first `fbsplash` works and shows the image on the display and the second call to `fbsplash` crashes?

**Comment by z3ntu** [comment]
Yes

**Comment by drebrez** [comment]
can you manually try to kill `msm-fb-refresher` right before calling `fbsplash` and start it right after?
Looking at the fbsplash code, when it open the framebuffer it does some ioctl call ([see here](https://github.com/mozilla-b2g/busybox/blob/master/miscutils/fbsplash.c#L87)) and since also `msm-fb-refresher` is doing stuff they go in conflict?

**Comment by z3ntu** [comment]
I tried
```
diff --git a/aports/main/postmarketos-mkinitfs/init_functions.sh b/aports/main/postmarketos-mkinitfs/init_functions.sh
index dc1f4c5..8853f55 100644
--- a/aports/main/postmarketos-mkinitfs/init_functions.sh
+++ b/aports/main/postmarketos-mkinitfs/init_functions.sh
@@ -293,7 +293,9 @@ start_usb_unlock() {
 # $1: path to ppm.gz file
 show_splash() {
        gzip -c -d "$1" >/tmp/splash.ppm
+       killall msm-fb-refresher 2>/dev/null
        fbsplash -s /tmp/splash.ppm
+       /usr/sbin/msm-fb-refresher --loop &
 }
 
 echo_connect_ssh_message() {
```
but the splash screen never showed. What worked though is that weston automatically shows and msm-fb-refresher is already running :) ("unintended side effect")

**Comment by ollieparanoid** [comment]
Good that you have tried it out! When searching for `mutex BUG: sleeping function called from invalid context`, a lot of results come up. It seems to be a bug in the kernel driver.

Luckily you are a kernel developer, so maybe you could figure out from there what goes wrong?

---

I didn't answer this one yet. @drebrez wrote:
> If we need 2 different solution we can consider having a `msm-fb-refresher` variant? what you think @ollieparanoid ?

If we needed two variants, I'd prefer if we extended `msm-fb-refresher` to handle both cases, maybe specify which one we need in the `deviceinfo` (as the `msm-fb-refresher` is from AsteroidOS, and upstream would benefit if we extended it).

**Comment by z3ntu** [comment]
No, I'm not a kernel developer 😄 I just did similar changes to the dts file I submitted upstream that I found with other devices (hammerhead and honami) ^^ And I got a lot of help on the irc channel 

So in short: I have no idea.

**Comment by ollieparanoid** [comment]
Okay then, cool that you are doing this anyway!

Is there another kernel you could try out, which potentially fixes this bug?

@pavelmachek: do you happen to know how to fix this?

**Comment by z3ntu** [comment]
No, only 3.4 (stock) and kinda-mainline-with-basically-nothing-working are available

**Comment by pavelmachek** [comment]
@ollieparanoid "sleeping function called from invalid context" means that you called something that can sleep from interrupt handler... probably. I'd have to understand the code. Fix may involve schedule_work() or something similar. https://www.safaribooksonline.com/library/view/understanding-the-linux/0596005652/ch04s08.html

**Comment by ollieparanoid** [comment]
@pavelmachek: thanks for the explanation!
@z3ntu: Actually the kernel [has been rebased](https://github.com/LineageOS/android_kernel_fairphone_msm8974/commit/2e055ec4343c34c11139b9d4568eaa322005603d) from `3.4.0` (according to the current APKBUILD) to `3.4.113`. There's a huge diff, so maybe this issue is fixed with the newer kernel. Do you want to give it a try?

**Comment by z3ntu** [comment]
That kernel has major stability problems, where there was (in my testing) some issue between 3.4.30 and 3.4.40. I have actually not yet tried the updated kernel with msm-fb-refresher, I'll try that. I am not expecting anything to change though as the problem probably lies within Qualcomm code which doesn't get fixed by the linux-stable tree.

**Comment by pavelmachek** [comment]
Agreed, it is unlikely that -stable kernel will fix anything
w.r.t. msm-fb-refresher.

On Sat 2017-10-07 12:47:24, Luca Weiss wrote:
> That kernel has major stability problems, where there was (in my testing) some issue between 3.4.30 and 3.4.40. I have actually not yet tried the updated kernel with msm-fb-refresher, I'll try that. I am not expecting anything to change though as the problem probably lies within Qualcomm code which doesn't get fixed by the linux-stable tree.
> 

-- 
(english) http://www.livejournal.com/~pavelmachek
(cesky, pictures) http://atrey.karlin.mff.cuni.cz/~pavel/picture/horses/blog.html

**Comment by montvid** [comment]
Well the msm-fb-refresher thing got fixed somewhere in the kernel as my Nexus 7 does not need it anymore on 4.11. And 3.4.113 is a bugfix release so it should be prefered to 3.4.0. Maybe it is a good idea to ask the linaro/msm guys on irc.

**Comment by z3ntu** [comment]
That's to be expected as mainline doesn't use the same drivers as 3.4. I am currently in the process of getting my panel to work on mainline but I have a few issues right now...
